build_tools/cmake/iree_check_test.cmake

```diff
@@ -268,10 +282,52 @@ function(iree_check_single_backend_test_suite)
         ${_RULE_OPT_TOOL}
       OPT_FLAGS
         ${_RULE_OPT_FLAGS}
+      TARGET_CPU_FEATURES
+        ${_RULE_TARGET_CPU_FEATURES}
     )
   endforeach()
 endfunction()
 
+# Helper function parsing a string occurring as an entry in TARGET_CPU_FEATURES_VARIANTS.
+#
+# This function has 3 output-params: variables that it sets with PARENT_SCOPE:
+# _ENABLED, _TARGET_CPU_FEATURES, _TARGET_CPU_FEATURES_SUFFIX.
+#
+# The special value "default" is handled here, resulting in the test always being enabled, with no target CPU feature flags and no suffix.
+#
+# Other values are parsed as "arch:features", the parsed arch is matched with CMAKE_SYSTEM_PROCESSOR, the test is enabled
+# if and only if there is a match, and then the parsed features are passed as target CPU featues, and the test suffix
+# is set accordingly.
+function(process_target_cpu_features _INPUT_TARGET_CPU_FEATURES _ENABLED _TARGET_CPU_FEATURES _TARGET_CPU_FEATURES_SUFFIX)
+  if ("${_INPUT_TARGET_CPU_FEATURES}" STREQUAL "default")
+    set(_ENABLED "TRUE" PARENT_SCOPE)
+    set(_TARGET_CPU_FEATURES "" PARENT_SCOPE)
+    set(_TARGET_CPU_FEATURES_SUFFIX "" PARENT_SCOPE)
+    return()
+    message(FATAL_ERROR "never go here")
+  endif()
+  string(REGEX MATCHALL "[^:]+" _COMPONENTS "${_INPUT_TARGET_CPU_FEATURES}")
+  list(LENGTH _COMPONENTS _NUM_COMPONENTS)
+  if (NOT _NUM_COMPONENTS EQUAL 2)
+    message (SEND_ERROR "_INPUT_TARGET_CPU_FEATURES should be of the form _FILTER_ARCH:_TARGET_CPU_FEATURES. Got: ${_INPUT_TARGET_CPU_FEATURES}")
+    return()
+  endif()
+  # TARGET_CPU_FEATURES_VARIANT is of the form _FILTER_ARCH:_TARGET_CPU_FEATURE.
+  list(GET _COMPONENTS 0 _FILTER_ARCH)
+  list(GET _COMPONENTS 1 _TARGET_CPU_FEATURES)
+  if (_FILTER_ARCH STREQUAL CMAKE_SYSTEM_PROCESSOR)
+    set(_ENABLED "TRUE" PARENT_SCOPE)
+    set(_TARGET_CPU_FEATURES "${_TARGET_CPU_FEATURES}" PARENT_SCOPE)
+    # TODO: the logic to generate the suffix from the list of target CPU features
```

Comment:
Can you add validation here to fail if any of these things are true, for now? regex match the whole string on `^[a-zA-Z0-9_+]+$` or something like that

Reply:
done